internal/envoy/auth/jwt_filter.go

```diff
@@ -0,0 +1,242 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package auth
+
+import (
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	envoy_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_config_route_v3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	envoy_jwt_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/jwt_authn/v3"
+
+	v1 "k8s.io/api/core/v1"
+
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/kubeshop/kusk-gateway/pkg/options"
+)
+
+// See https://github.com/projectcontour/contour/blob/main/internal/envoy/v3/listener.go#L746 for example usage of the filter.
+
+func NewFilterHTTPJWT(jwtOptions *options.JWT, args *ParseAuthArguments, paths []string) (*anypb.Any, error) {
+	if len(jwtOptions.JWTProviders) == 0 {
+		return nil, nil
+	}
+
+	jwtConfig := envoy_jwt_v3.JwtAuthentication{
+		Providers:      map[string]*envoy_jwt_v3.JwtProvider{},
+		RequirementMap: map[string]*envoy_jwt_v3.JwtRequirement{},
+		Rules:          []*envoy_jwt_v3.RequirementRule{},
+	}
+
+	for _, provider := range jwtOptions.JWTProviders {
+		// Default cache timeout.
+		var cacheDuration *durationpb.Duration = durationpb.New(time.Second * 300)
+		if provider.RemoteJWKS.CacheDuration != nil {
+			cacheDuration = durationpb.New(*provider.RemoteJWKS.CacheDuration)
+		}
+		// Default Remote JWKS timeout.
+		var timeout *durationpb.Duration = durationpb.New(time.Second * 1)
+		if provider.RemoteJWKS.Timeout != nil {
+			timeout = durationpb.New(*provider.RemoteJWKS.Timeout)
+		}
+
+		cluster := provider.RemoteJWKS.URI
+		if !args.EnvoyConfiguration.ClusterExist(cluster) {
+			url, err := url.Parse(cluster)
+			if err != nil {
+				return nil, err
+			}
+
+			upstreamServiceHost := url.Hostname()
+			upstreamServicePort := uint32(443)
+
+			port := url.Port()
+			if port != "" {
```

Comment:
what happens if port == ""?

Reply:
An example of when port is `""` is when the `url` parsed is `https://kubeshop-kusk-gateway-oauth2.eu.auth0.com/.well-known/jwks.json`.